protocol-units/da-sequencer/node/src/storage/mod.rs

```diff
@@ -1,74 +1,404 @@
-use crate::batch::DaBatch;
-use crate::batch::FullnodeTx;
-use crate::block::BlockHeight;
-use crate::block::SequencerBlock;
-use crate::block::SequencerBlockDigest;
-use crate::celestia::CelestiaHeight;
-use crate::error::DaSequencerError;
+use crate::{
+	batch::{DaBatch, FullNodeTx},
+	block::{BlockHeight, SequencerBlock, SequencerBlockDigest},
+	celestia::CelestiaHeight,
+	error::DaSequencerError,
+};
+use bcs;
+use movement_types::block::{Block, BlockMetadata, Id};
+use movement_types::transaction::Transaction;
+use rocksdb::{ColumnFamilyDescriptor, Options, WriteBatch, DB};
+use std::{collections::BTreeSet, result::Result, sync::Arc};
 
-pub struct Storage {}
+pub mod cf {
+	pub const PENDING_TRANSACTIONS: &str = "pending_transactions";
+	pub const BLOCKS: &str = "blocks";
+	pub const BLOCKS_BY_DIGEST: &str = "blocks_by_digest";
+}
+
+#[derive(Debug)]
+pub struct Storage {
+	db: Arc<DB>,
+}
 
 impl Storage {
-	/// Save all batch's Tx in the pending Tx table. The batch's Tx has been verified and validated.
-	pub fn write_batch(
-		&self,
-		batch: DaBatch<FullnodeTx>,
-	) -> std::result::Result<(), DaSequencerError> {
-		todo!();
+	pub fn try_new(path: &str) -> Result<Self, DaSequencerError> {
+		let mut options = Options::default();
+
+		options.create_if_missing(true);
+		options.create_missing_column_families(true);
+
+		let pending_transactions_cf =
+			ColumnFamilyDescriptor::new(cf::PENDING_TRANSACTIONS, Options::default());
+		let blocks_cf = ColumnFamilyDescriptor::new(cf::BLOCKS, Options::default());
+		let blocks_by_digest_cf =
+			ColumnFamilyDescriptor::new(cf::BLOCKS_BY_DIGEST, Options::default());
+
+		let db = DB::open_cf_descriptors(
+			&options,
+			path,
+			[pending_transactions_cf, blocks_cf, blocks_by_digest_cf],
+		)
+		.map_err(|e| DaSequencerError::Generic(e.to_string()))?;
+
+		Ok(Storage { db: Arc::new(db) })
+	}
+
+	pub fn write_batch(&self, batch: DaBatch<FullNodeTx>) -> Result<(), DaSequencerError> {
+		let cf = self.db.cf_handle(cf::PENDING_TRANSACTIONS).ok_or_else(|| {
+			DaSequencerError::Generic("Missing column family: pending_transactions".into())
+		})?;
+		let tx = batch.data();
+		let key = tx.id.0;
+		let value = bcs::to_bytes(tx)
```

Comment:
We use the Tx batch to build a block. In the worst case, not all tx's batch can fit in the block, so you've to fill the block and update the batch in this case. Perhaps it's easier to save each Tx individually with the batch timestamp, in this case when the Tx are added to a saved block you need to delete the Tx.

Reply:
To be addressed in followup `produce_block` PR.